nixos/modules/services/misc/nix-daemon.nix

```
@@ -383,8 +383,15 @@ in
         type = types.bool;
         default = true;
         description = lib.mdDoc ''
-          If enabled (the default), checks for data type mismatches and that Nix
-          can parse the generated nix.conf.
+          If enabled (the default), checks that Nix can parse the generated nix.conf.
+        '';
+      };
+
+      checkAllErrors = mkOption {
+        type = types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          If enabled (the default), checks the nix.conf parsing for any kind of error. When disabled, only unknown settings are checked for.
         '';
       };
```

```suggestion
      checkAllErrors = mkOption {
        type = types.bool;
        default = true;
        description = lib.mdDoc ''
          If enabled (the default), checks the nix.conf parsing for any kind of error. When disabled, checks only for unknown settings.
        '';
      };
```
Maybe add a scenario when this is helpful to make it easier for people to know when to use this?

---

I don't actually know any scenario where you might want this but that's how the option worked before, so I retained that functionality.

---

If we don't know any usecase for this setting then why do we have it at all? Just because it was always like does not mean we can't change it.

---

Sure, if you're motivated to find out, go ahead.

I'm not, so I opted to just retain the old behaviour as an option. It's also not like this option really costs us anything. It's not very complex and might help someone's odd use-case.

---

The purpose of having a check at the derivation level for unknown settings is to emulate checks for attribute names at the module system level. For instance, if every nix.conf setting is transformed to a corresponding typed module option, then

```nix
...
nix.settings = {
  some-unknown-setting = true;
};
...
```

would produce an error on evaluation because `settings.some-unknown-setting` is not defined. However, the current implementation would happily generate the nix.conf with the setting.

The benefit of freeform modules versus mapping all settings to module options is that it is easier to maintain but it comes at a cost of having less stringent type checking at evaluation time.

The reason the intermediate state of checking for unknown settings is needed is because `nix show-config` isn't hermetic. One case I recall where it is necessary to relax the strictness of the check is if the system is not connected to the network, `nix show-config` would output "warning: you don't have Internet access; disabling some network-dependent features" which gets converted to an error, blocking the derivation build.